File: JeongSH/Programmers/Greedy/MakeMax.java

import java.util.Stack;
/*
    문제 아이디어:
    number 값 중 앞에서 부터 작은 수를 지우는게 큰 수를 만드는 방법이라고 생각했습니다.

    시간복잡도: O(n^2)

 */


public class MakeMax {
    public static void main(String args[]){
        String number = "987654321";
        int k=4;
        char[] result = new char[number.length()-k];
        Stack<Character> stack = new Stack<>();
        for(int i=0; i< number.length(); i++){
            char c= number.charAt(i);

            while(!stack.empty() && stack.peek()<c && k-- > 0){     // stack의 최상단을 확인하여 i번 째의 수가 가장 큰 수인지 확인합니다.
                stack.pop();                                        // stack의 최상단이 i보다 작다면 pop을 해줍니다. 최대 k번 까지.
            }
            stack.push(c);
        }
        for(int i=0; i<result.length; i++){
            result[i]= stack.get(i);                                // 앞에서 작은 숫자를 지웠으나 k번 보다 덜 지워졌을 경우
        }                                                           // ex) 987654321 같은 경우 뒤에서 남은 k번 만큼 지워야합니다.
        System.out.println(new String(result));
    }
}

File: 6loss0m/Programmers/Greedy/makeMax.py
def solution(number, k):
    stack = [number[0]]
    for num in number[1:]:
        # 1. 스택이 채워져 있음
        # 2. top이 현재의 수보다 작음
        # 3. 제외할 숫자가 아직 있음
        while len(stack) > 0 and stack[-1] < num and k > 0:
            k -= 1
            stack.pop()
        stack.append(num)
    print(stack)
    if k != 0:
        stack = stack[:-k]
    return ''.join(stack)